core/crypto/vrf_provider.hpp

```diff
@@ -0,0 +1,44 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_CORE_CRYPTO_VRF_VRF_PROVIDER_HPP
+#define KAGOME_CORE_CRYPTO_VRF_VRF_PROVIDER_HPP
+
+#include "common/buffer.hpp"
+#include "crypto/vrf_types.hpp"
+
+namespace kagome::crypto {
+
+  /**
+   * SR25519 based verifiable random function implementation
+   */
+  class VRFProvider {
+   public:
+    virtual ~VRFProvider() = default;
+
+    /**
+     * Generates random keypair for signing the message
+     */
+    virtual VRFKeypair generateKeypair() const = 0;
+
+    /**
+     * Sign message \param msg using \param keypair. If computed value is less
+     * than \param threshold then return optional containing this value and
+     * proof. Otherwise none returned
+     */
+    virtual boost::optional<VRFOutput> signMsg(
+        const common::Buffer &msg, const VRFKeypair &keypair,
+        const VRFValue &threshold) const = 0;
+
+    /**
+     * Verifies that \param output was derived using \param public_key on \param
+     * msg
+     */
+    virtual bool verifyMsg(const common::Buffer &msg, const VRFOutput &output,
```

Comment:
Not sure if it a good idea to use common::Buffer, since it is an additional dependency.

Reply:
message will be composed by concatenating several structures, so buffer fits perfectly

Reply:
It seems to be necessary, ok.